lib/query/index.js

```diff
@@ -0,0 +1,129 @@
+const _ = require('lodash')
+const { filterAndTransform } = require('./filter-and-transform')
+const { logWarnings } = require('./log-warnings')
+const { renderFeaturesResponse } = require('./render-features')
+const { renderStatisticsResponse } = require('./render-statistics')
+const { renderPrecalulatedStatisticsResponse } = require('./render-precalculated-statistics')
+const { renderCountAndExtentResponse } = require('./render-count-and-exent')
+const { getGeometryTypeFromGeojson } = require('../helpers')
+
+function query (json, requestParams = {}) {
+  const {
+    features,
+    filtersApplied: {
+      all: skipFiltering
+    } = {}
+  } = json
+
+  const { f: requestedFormat } = requestParams
+
+  if (shouldRenderPrecalculatedData(json, requestParams)) {
+    return renderPrecalculatedData(json, requestParams)
+  }
+
+  const data = (skipFiltering || !features) ? json : filterAndTransform(json, requestParams)
+
+  if (shouldLogWarnings()) {
+    logWarnings(data, requestParams.f)
+  }
+
+  if (requestedFormat === 'geojson') {
+    return {
+      type: 'FeatureCollection',
+      features: data.features
+    }
+  }
+
+  return renderGeoservicesResponse(data, {
+    ...requestParams,
+    attributeSample: _.get(json, 'features[0].properties'),
+    geometryType: getGeometryTypeFromGeojson(json)
+  })
+}
+
+function shouldRenderPrecalculatedData ({ statistics, count, extent }, { returnCountOnly, returnExtentOnly }) {
+  if (statistics) {
+    return true
+  }
+
+  if (returnCountOnly === true && count !== undefined && returnExtentOnly === true && extent) {
+    return true
+  }
+
+  if (returnCountOnly === true && count !== undefined && !returnExtentOnly) {
+    return true
+  }
+
+  if (returnExtentOnly === true && extent && !returnCountOnly) {
+    return true
+  }
+
+  return false
+}
+
+function renderPrecalculatedData (data, { returnCountOnly, returnExtentOnly }) {
+  const { statistics, count, extent } = data
+
+  if (statistics) {
+    return renderPrecalulatedStatisticsResponse(data)
+  }
+
+  const retVal = {}
+
+  if (returnCountOnly) {
+    retVal.count = count
+  }
+
+  if (returnExtentOnly) {
+    retVal.extent = extent
+  }
```

Comment:
This part's a bit confusing cause of the variable names. I see `returnCountOnly` and expect `renderPrecalculatedData()` to just return `count`. But then it goes on and checks `returnExtentOnly` and also returns `extent`. I would've expected just one or the other but maybe it's the lack of context here?

Reply:
Yes, it's weird.  But it's part of the GeoServices API spec.  You can make a request with params `returnCountOnly` and `returnExtentOnly` and it will return both count and extent despite the `Only` references.  Nothing we can do if we have to honor GeoServices API spec.  All the weirdness documented here: https://developers.arcgis.com/rest/services-reference/enterprise/query-feature-service-layer-.htm